src/cmd/apply.rs

```diff
@@ -0,0 +1,152 @@
+use CliResult;
+use CliError;
+use config::{Config, Delimiter};
+use select::SelectColumns;
+use util;
+
+use std::collections::VecDeque;
+use std::io::{Write, BufReader, BufRead};
+use std::process;
+use std::sync::mpsc;
+use std::thread;
+
+static USAGE: &'static str = "
+Runs an external program over specified columns of a csv file.
+
+Usage:
+    xsv apply <column-name> <executable> [options] [<input>]
+
+The program will be started in a subshell, the column will be sent to stdin of the process and we
+will replace the existing column with stdout of the process.
+
+TODO flag to use index instead of name
+
+Common options:
+    -h, --help             Display this message
+    -o, --output <file>    Write output to <file> instead of stdout.
+    -d, --delimiter <arg>  The field delimiter for reading CSV data.
+                           Must be a single character. (default: ,)
+";
+
+#[derive(RustcDecodable)]
+struct Args {
+    arg_column_name: SelectColumns,
+    arg_executable: String,
+    arg_input: Option<String>,
+    flag_output: Option<String>,
+    flag_delimiter: Option<Delimiter>,
+}
+
+pub fn run(argv: &[&str]) -> CliResult<()> {
+    let args: Args = util::get_args(USAGE, argv)?;
+
+    let rconfig = Config::new(&args.arg_input)
+                         .delimiter(args.flag_delimiter)
+                         .select(args.arg_column_name);
+
+    let mut rdr = rconfig.reader()?;
+    let mut wtr = Config::new(&args.flag_output).writer()?;
+
+    let headers = rdr.byte_headers()?;
+    let sel = rconfig.selection(&*headers)?;
+
+    // start up the program
+    let prog = process::Command::new("sh")
+        .arg("-c")
+        .arg(args.arg_executable)
+        .stdout(process::Stdio::piped())
+        .stdin(process::Stdio::piped())
+        .spawn();
+
+    let mut prog = match prog {
+        Ok(p)  => p,
+        Err(e) => return Err(CliError::Io(e))
+    };
+
+    let progout = match prog.stdout.take() {
+        Some(i) => BufReader::new(i),
+        None    => panic!("return a better error here")
+    };
+
+    // create a thread to keep up with the external program's output
+    // let these lines get buffered in the channel
```

Comment:
If we don't consume from the pipe, the external program will (probably) hang. Then, we (probably) won't be able to write to the child's stdin (if they stop consuming from the pipe, the buffer will fill up), and everyone get's stuck.

If we had some non-blocking io in `std::io` I could attempt a read at the bottom of the loop on starting line 95. If the read were successful, then I could read as many lines as possible from the read buffer. If there were no data ready to be ready, I could just keep pushing lines to child's stdin.

We don't have this kind of support in the standard library, so I've created the same behavior with a channel (because those can be "non-blocking" in a sense).

I don't like it, but it seemed like this was the path of least resistance.